src/main/java/com/limechain/trie/structure/decoded/node/DecodedNode.java

```diff
@@ -228,4 +240,221 @@ public byte[] calculateMerkleValue(Function<byte[], byte[]> hashFunction, boolea
+     *
+     * @param reader             ScaleCodecReader used to read the encoded node data.
+     * @param firstByte          The first byte of the encoded node data.
+     * @param pkLenFirstByteBits The number of bits in the first byte dedicated to encoding the partial key length.
+     * @return int The length of the partial key.
+     * @throws NodeDecodingException If there's an error in decoding, such as if the partial key length is too short,
+     *                               or if there's an overflow in calculating the partial key length.
+     */
+    private static int getPkLen(ScaleCodecReader reader, int firstByte, int pkLenFirstByteBits) {
+        int pkLen = firstByte & ((1 << pkLenFirstByteBits) - 1);
+
+        boolean continueIter = pkLen == ((1 << pkLenFirstByteBits) - 1);
+
+        while (continueIter) {
+            if (!reader.hasNext()) {
+                throw new NodeDecodingException("PartialKeyLenTooShort");
+            }
+            byte currentByte = reader.readByte();
+            continueIter = (currentByte & 0xFF) == 255;
+            int addedValue = (currentByte & 0xFF);
+
+            if (pkLen > Integer.MAX_VALUE - addedValue) {
+                throw new NodeDecodingException("Partial key length overflow");
+            }
+
+            pkLen += addedValue;
+        }
+        return pkLen;
+    }
+
+    /**
+     * Extracts the partial key from a node in a trie structure using a ScaleCodecReader.
+     *
+     * @param reader The ScaleCodecReader used to read the encoded node data.
+     * @param pkLen  The length of the partial key in nibbles.
+     * @return byte[] The byte array containing the partial key.
+     * @throws NodeDecodingException If there are no more bytes to read, or if the partial key has invalid padding.
+     */
+    public static byte[] extractPartialKey(ScaleCodecReader reader, int pkLen) {
+        if (!reader.hasNext()) {
+            throw new NodeDecodingException("Node value cannot be null");
+        }
+
+        // Calculate the length of the partial key in bytes
+        int pkLenBytes = pkLen == 0 ? 0 : 1 + ((pkLen - 1) / 2);
+        byte[] partialKey = reader.readByteArray(pkLenBytes);
+
+        if ((pkLen % 2 == 1) && ((partialKey[0] & 0xf0) != 0)) {
+            throw new NodeDecodingException("Invalid Partial key padding");
+        }
+
+        return partialKey;
+    }
+
+    /**
+     * Extracts the children bitmap from a node in a trie structure.
+     *
+     * @param reader      The ScaleCodecReader used to read the encoded node data.
+     * @param hasChildren A boolean indicating whether the node has children.
+     * @return int The children bitmap as an integer. If the node has no children, returns 0.
+     * @throws NodeDecodingException If there are no more bytes to read, or if the children bitmap is zero.
+     */
+    public static int extractChildrenBitmap(ScaleCodecReader reader, boolean hasChildren) {
```

Comment:
Should this be public static ?

Reply:
Changed them to private but I'd like to leave them static